### ISSUE TITLE
extract-i18n produces file that breaks build (when two <a href=""> tags are embedded inside the translation)

### ISSUE THREAD

**Issue by mstawick** [angular/angular#41142]
# 🐞 Bug report

### Command (mark with an `x`)
- [ ] new
- [x] build
- [ ] serve
- [ ] test
- [ ] e2e
- [ ] generate
- [ ] add
- [ ] update
- [ ] lint
- [x] extract-i18n
- [ ] run
- [ ] config
- [ ] help
- [ ] version
- [ ] doc

Not sure if it's the extraction process or the build process

### Is this a regression?
I don't think so - I think it's been there since localize was introduced

### Description
Having a specific combination of tags within the translation unit will produce a translation file that breaks the build. Please see the reproduction section.

## 🔬 Minimal Reproduction
Since it has to do with localize and building the project, I'm attaching a zipped reproduction.

Having such a piece of HTML:
`<div class="some-class" i18n>This breaks build <a href="https://foo.com">One link</a> the second link causes error <a href="https://bar.com">Another link</a></div>`

Will produce such translate unit (xlf2, with target duplicated from source):
```   
    <unit id="4984752463945187369">
      <notes>
        <note category="location">src/app/app.component.html:1</note>
      </notes>
      <segment>
        <source>This breaks build <pc id="0" equivStart="START_LINK" equivEnd="CLOSE_LINK" type="link" dispStart="&lt;a href=&quot;https://foo.com&quot;&gt;" dispEnd="&lt;/a&gt;">One link</pc> the second link causes error <pc id="1" equivStart="START_LINK_1" equivEnd="CLOSE_LINK_1" type="other" dispStart="&lt;a href=&quot;https://bar.com&quot;&gt;">Another link</pc></source>
        <target>This breaks build <pc id="0" equivStart="START_LINK" equivEnd="CLOSE_LINK" type="link" dispStart="&lt;a href=&quot;https://foo.com&quot;&gt;" dispEnd="&lt;/a&gt;">One link</pc> the second link causes error <pc id="1" equivStart="START_LINK_1" equivEnd="CLOSE_LINK_1" type="other" dispStart="&lt;a href=&quot;https://bar.com&quot;&gt;">Another link</pc></target>
      </segment>
    </unit>
```

which then causes the build to fail (ng build --localize) with error message:
`There is a placeholder name mismatch with the translation provided for the message "8220418496226875748" ["ae049955b00b7c32e119cf04537cb0de7609cdb2", "4984752463945187369"] ("This breaks build {$START_LINK}One link{$CLOSE_LINK} the second link causes error {$START_LINK_1}Another link{$CLOSE_LINK}").
The translation contains a placeholder with name CLOSE_LINK_1, which does not exist in the message.`

When tested with bare tags: `<a>Text</a>`, without the href attribute, it worked fine.

## 🔥 Exception or Error
<pre><code>
There is a placeholder name mismatch with the translation provided for the message "8220418496226875748" ["ae049955b00b7c32e119cf04537cb0de7609cdb2", "4984752463945187369"] ("This breaks build {$START_LINK}One link{$CLOSE_LINK} the second link causes error {$START_LINK_1}Another link{$CLOSE_LINK}").
The translation contains a placeholder with name CLOSE_LINK_1, which does not exist in the message.
</code></pre>

## 🌍 Your Environment
<pre><code>
Angular CLI: 11.0.7
Node: 14.14.0
OS: linux x64

Angular: 11.0.9
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.7
@angular-devkit/build-angular   0.1100.7
@angular-devkit/core            11.0.7
@angular-devkit/schematics      11.0.7
@angular/cli                    11.0.7
@schematics/angular             11.0.7
@schematics/update              0.1100.7
rxjs                            6.6.6
typescript                      4.0.7
</code></pre>
[i18n-tags-bug.tar.gz](https://github.com/angular/angular-cli/files/6111085/i18n-tags-bug.tar.gz)

**Comment by petebacondarwin** [comment]
Thanks for posting this issue. I'll look into it tomorrow.

**Comment by walli55** [comment]
https://Iecdf4cb7f6815f65e7abceb85b61e8d85d80fa9e

**Comment by petebacondarwin** [comment]
Note to self: the second link is being rendered in the XLIFF as `<pc type="other">`, with no `dispEnd` attribute, rather than `type="link"`.

**Comment by trojanhorsekilla** [comment]
I have lots to learn. It have to fully understand everything you mentioned .My iOS is compromised I know my clipboard altered I’m getting a third-party keyboard

**Comment by petebacondarwin** [comment]
Next note to self... the problem is that in the template compiler we are assuming that the first and second `</a>` closing tags can be considered equal, while the XLIFF serializer does not.

We could hack the serializer to workaround this, but I think the correct solution is to repair the template compiler. Unfortunately, this change would change the generated message IDs for such messages, which would be a breaking change, and not so easy to automatically migrate.

Perhaps we could mitigate the breaking change by avoiding a change to the legacy message IDs, which are due to be migrated anyway. This would result in only affecting users who had turned off legacy message IDs in their projects, which includes all newly created CLI projects.

**Comment by petebacondarwin** [comment]
Actually, I think the hacky solution will be adequate, since we only use source-mapping to work out the `dispEnd` value, which will be the same for all of the closing tags.

**Comment by AndrewKushnir** [comment]
@mstawick thanks for reporting the issue. Just want to let you know that the fix created by Pete was merged into main branches and the change should be available in the patch release (v11.2.6) next week. Thank you.

**Comment by mstawick** [comment]
@AndrewKushnir @petebacondarwin Thank you guys!
